File: src/main/java/refactown/cleancode/c07exceptions/java/JsonParsing.java
package refactown.cleancode.c07exceptions.java;

import com.google.gson.Gson;

public class JsonParsing {

	public static Object fromJsonV0(String json, Class<?> clazz) {
		Gson gson = new Gson();
		try {
			return gson.fromJson(json, clazz);
		} catch (Exception e) {
			System.out.println("Parsing error. Json text:" + json);
			return null;
		}
	}

	public static Object fromJsonV1(String json, Class<?> clazz) {
		Gson gson = new Gson();
		try {
			return gson.fromJson(json, clazz);
		} catch (Exception e) {
			throw new JsonException("Parsing error. Json text:$json");
		}
	}

	public static Object fromJsonV2(String json, Class<?> clazz) {
		if (json == null){
			throw new NullPointerException("A variável json não pode ser nula!");
		}
		Gson gson = new Gson();
		try {
			return gson.fromJson(json, clazz);
		} catch (Exception e) {
			throw new JsonException("Parsing error. Json text:$json", e); // CAUSE
		}
	}

}

File: src/main/kotlin/refactown/cleancode/c07exceptions/JsonParsing.kt
package refactown.cleancode.c07exceptions

import com.google.gson.Gson

typealias JsonParserFunction = (String, Class<*>) -> Any

fun fromJsonV0(json: String, clazz: Class<*>): Any?{
    val gson = Gson()
    try {
        return gson.fromJson(json, clazz)
    } catch (e: Exception){
        println("Parsing error. Json text:$json")
        return null
    }
}

fun fromJsonV1(json: String, clazz: Class<*>): Any{
    val gson = Gson()
    try {
        return gson.fromJson(json, clazz)
    } catch (e: Exception){
        throw JsonException("Parsing error. Json text:$json")
    }
}

fun fromJsonV2(json: String, clazz: Class<*>): Any{
    val gson = Gson()
    try {
        return gson.fromJson(json, clazz)
    } catch (e: Exception){
        throw JsonException("Parsing error. Json text:$json", e)
    }
}
